src/dragonball/src/address_space_manager.rs

```diff
@@ -0,0 +1,829 @@
+            // will pre-fault in all memory with normal pages before madive(MADV_HUGEPAGE) gets
+            // called. So remove the MAP_POPULATE flag and memory will be faulted in by working
+            // threads.
+            region.perm_flags() & (!libc::MAP_POPULATE)
+        } else {
+            region.perm_flags()
+        };
+        let mmap_reg = MmapRegion::build(
+            file_offset,
+            region.len() as usize,
+            libc::PROT_READ | libc::PROT_WRITE,
+            perm_flags,
+        )
+        .map_err(AddressManagerError::MmapGuestMemory)?;
+
+        if region.is_anonpage() {
+            self.configure_anon_mem(&mmap_reg)?;
+        }
+        if let Some(node_id) = region.host_numa_node_id() {
+            self.configure_numa(&mmap_reg, node_id)?;
+        }
+        if region.is_hugepage() {
+            self.configure_thp_and_prealloc(&region, &mmap_reg)?;
+        }
+
+        let reg = GuestRegionImpl::new(mmap_reg, region.start_addr())
+            .map_err(AddressManagerError::CreateGuestMemory)?;
+        Ok(Arc::new(reg))
+    }
+
+    fn configure_anon_mem(&self, mmap_reg: &MmapRegion) -> Result<()> {
+        unsafe {
+            mman::madvise(
+                mmap_reg.as_ptr() as *mut libc::c_void,
+                mmap_reg.size(),
+                mman::MmapAdvise::MADV_DONTFORK,
+            )
+        }
+        .map_err(AddressManagerError::Madvise)
+    }
+
+    fn configure_numa(&self, mmap_reg: &MmapRegion, node_id: u32) -> Result<()> {
+        let nodemask = (1 << node_id) as u64;
+        let res = unsafe {
+            libc::syscall(
+                libc::SYS_mbind,
+                mmap_reg.as_ptr() as *mut libc::c_void,
+                mmap_reg.size(),
+                MPOL_PREFERRED,
+                &nodemask as *const u64,
+                MAX_NODE,
+                MPOL_MF_MOVE,
+            )
+        };
+        if res < 0 {
+            warn!("failed to mbind memory to host_numa_node_id {}", node_id);
+        }
+        Ok(())
+    }
+
+    fn configure_thp_and_prealloc(
+        &mut self,
+        region: &Arc<AddressSpaceRegion>,
+        mmap_reg: &MmapRegion,
+    ) -> Result<()> {
+        debug!(
+            "Setting MADV_HUGEPAGE on AddressSpaceRegion addr {:x?} len {:x?}",
+            mmap_reg.as_ptr(),
+            mmap_reg.size()
+        );
+
+        // Safe because we just create the MmapRegion
+        unsafe {
+            mman::madvise(
+                mmap_reg.as_ptr() as *mut libc::c_void,
+                mmap_reg.size(),
+                mman::MmapAdvise::MADV_HUGEPAGE,
+            )
+        }
+        .map_err(AddressManagerError::Madvise)?;
+
+        // Touch every 4k page to trigger allocation. The step is 4K instead of 2M to ensure
+        // pre-allocation when running out of huge pages.
+        if region.perm_flags() & libc::MAP_POPULATE > 0 {
+            const PAGE_SIZE: u64 = 4096;
```

Comment:
Because the memory pre-allocation process only wants to touch every 4K page instead of 2M or other page size settings to avoid running out of huge pages so we fix that const to 4096